controllers/tls_workflow.go

```diff
@@ -133,24 +126,126 @@ func LinkGatewayToClusterIssuerFunc(objs controller.Store) machinery.LinkFunc {
 				return p.Spec.IssuerRef.Name == clusterIssuer.GetName() && p.Spec.IssuerRef.Kind == certmanagerv1.ClusterIssuerKind
 			})
 
-			if len(linkedPolicies) == 0 {
-				return nil
+			return findLinkedGatewaysForIssuer(linkedPolicies, gateways)
+		},
+	}
+}
+
+func findLinkedGatewaysForIssuer(linkedPolicies []*kuadrantv1alpha1.TLSPolicy, gateways []*gwapiv1.Gateway) []machinery.Object {
+	if len(linkedPolicies) == 0 {
+		return nil
+	}
+
+	// Can infer linked gateways through the policy
+	linkedGateways := lo.Filter(gateways, func(g *gwapiv1.Gateway, index int) bool {
+		for _, l := range linkedPolicies {
+			if string(l.Spec.TargetRef.Name) == g.GetName() && g.GetNamespace() == l.GetNamespace() {
+				return true
+			}
+		}
+
+		return false
+	})
+
+	return lo.Map(linkedGateways, func(item *gwapiv1.Gateway, index int) machinery.Object {
+		return &machinery.Gateway{Gateway: item}
+	})
+}
+
+// Common functions used across multiple reconcilers
+
+func IsTLSPolicyValid(ctx context.Context, s *sync.Map, policy *kuadrantv1alpha1.TLSPolicy) (bool, error) {
+	logger := controller.LoggerFromContext(ctx).WithName("IsPolicyValid")
+
+	store, ok := s.Load(TLSPolicyAcceptedKey)
+	if !ok {
+		logger.V(1).Info("TLSPolicyAcceptedKey not found, policies will be checked for validity by current status")
+		return meta.IsStatusConditionTrue(policy.Status.Conditions, string(gatewayapiv1alpha2.PolicyReasonAccepted)), nil
+	}
+
+	isPolicyValidErrorMap := store.(map[string]error)
+
+	return isPolicyValidErrorMap[policy.GetLocator()] == nil, isPolicyValidErrorMap[policy.GetLocator()]
+}
+
+func GetTLSPoliciesByEvents(topology *machinery.Topology, events []controller.ResourceEvent) []machinery.Policy {
+	policies := lo.Filter(topology.Policies().Items(), func(item machinery.Policy, index int) bool {
+		_, ok := item.(*kuadrantv1alpha1.TLSPolicy)
+		return ok
+	})
+
+	var affectedPolicies []machinery.Policy
+	for _, event := range events {
+		if event.Kind == machinery.GatewayGroupKind {
```

Comment:
I am not sure how I feel about looping over the events like this. However, would it not make more sense to use a  switch statement over this individual if blocks,

Reply:
I've reverted this logic so maybe we can revisit this again at a later date if we want to re-introduce it